### ISSUE TITLE
Missing translations callback&fallback

### ISSUE THREAD

**Issue by cranetm** [ryandrewjohnson/react-localize-redux#36]
If some translation is missing, I can not setup returned value. There is _showMissingTranslationMsg_ and _missingTranslationCallback_, but i really want have some power to make my own output string.
As possible solution you could return _missingTranslationCallback_ result value if callback exists (utils.js, line 9).
And fallback part. It will be really usefully, if translation for some language is missing then return (or try to find) it from default language.
What do you think?

**Comment by ryandrewjohnson** [comment]
@cranetm I have a few questions to better understand the feature request.

So the way things currently work is as follows:

Let's say you have translation data that doesn't contain a key:
```javascript
// this key doesn't exist
<h1>{translate('title')}</h1>

// so it will render the following
<h1>Missing localized key: title for language: en</h1>
```
Now this would occur for one of two reasons:

1. The translation data for `title` doesn't exist at all.
2. The translation data for `title` exists, but is missing for the current language.

I could add an `initialize` option that would allow you to set a custom string for missing translation. For example maybe you might want to use an empty string for production, so you'd never get the `Missing localized key:` message in prod.

Would this cover your use case or are you requirements different?

**Comment by cranetm** [comment]
Ok, i will describe detailed.
We do not have title in translation:

`<h1>{translate('title')}</h1>`

now it renders "Missing localized key..." message if showMissingTranslationMsg is true.
I propose to render result of missingTranslationCallback if its defined. And I can manipulate with it in any way I want:
```
const missingTranslationCallback = (key: string, languageCode: string) => {
  InternalLog.warn('Missing key "' + key + '" for "' + languageCode '".');
  return key;
  // or transform to more user friendly
  return key.charAt(0).toUpperCase() + key.slice(1);
};
```

And second is about fallback to default language.
For example, all development doing in english and with a high probability all english translations present. Application has 3 more languages, which will be added in some future by technical writers.
To prevent empty strings or unnecessary messages, event in development stage, it can render translation from english.

Actually you can combine both. Add third params to missingTranslationCallback, called like _defaultTranslation: string_ and this value could be returned and rendered if translation is missing.

Is is clear enough?

**Comment by ryandrewjohnson** [comment]
@cranetm first off thanks for the suggestions. I've included my feedback below.
 
The requested changes for `missingTranslationCallback` seem very specific to your use case. In addition it would introduce breaking changes to the API, which I wouldn't want to make for a change this small.

In terms of the fallback to default language, this is something I'll take a look at. My initial thought is to add a new option to `initialize` to allow user to opt in to this feature. I will update this issue as things progress on the work, or if you'd like to contribute feel free to submit a PR.

**Comment by MaiGhoneim** [comment]
hi, 
missingTranslationCallback did not change missing localization message, i need to change this message how i can do it

**Comment by ryandrewjohnson** [comment]
@MaiGhoneim the missingTranslationCallback is just a function that will trigger when a translation is missing. It does not actually change the value of the missing translation message. You can set showMissingTranslationMsg to false to hide the missing translation message, but there isn’t currently a way to define a custom missing message.

If you want open a new issue and label it feature request, and explain the feature and use case clearly  I can have a look.

**Comment by vevmesteren** [comment]
Can one language be set as fall-back. When a translation missing I would want to display the English translation.

**Comment by ryandrewjohnson** [comment]
@vevmesteren I'm most likely going to table this feature until the next major version release. The reason being that there are already three [options](https://ryandrewjohnson.github.io/react-localize-redux/api/action-creators/#initializelanguages-options) dedicated to the missing translations:

* `showMissingTranslationMsg`
* `missingTranslationMsg`
* `missingTranslationCallback`

These options are already the result of squeezing in features, and as much as I'd like to rethink the whole missing translations implementation it would result in breaking changes - hence the waiting for a major version release. I'm targeting a major version release for in and around when React releases their new context API, as there are changes I want to make that are dependent on that.

**Comment by vevmesteren** [comment]
That'll do I'll figure something out with those. 
Thanks

**Comment by ryandrewjohnson** [comment]
I'm considering this closed with the release of [v3.0.0](https://github.com/ryandrewjohnson/react-localize-redux/releases/tag/v3.0.0). 

See [Handle missing translations](https://ryandrewjohnson.github.io/react-localize-redux-docs/#handle-missing-translations) in the new docs for details.
